py/misc.h

```diff
@@ -74,14 +74,21 @@ typedef unsigned int uint;
 
 // TODO make a lazy m_renew that can increase by a smaller amount than requested (but by at least 1 more element)
 
-#define m_new(type, num) ((type *)(m_malloc(sizeof(type) * (num))))
-#define m_new_maybe(type, num) ((type *)(m_malloc_maybe(sizeof(type) * (num))))
-#define m_new0(type, num) ((type *)(m_malloc0(sizeof(type) * (num))))
+// The following are convenience wrappers for m_malloc_helper and can save space at the call sites.
+// m_malloc and m_new allocate space that is not collected and does not have a finaliser.
+// Use m_malloc_items() to allocate space for mp_obj_t that will be collected.
+// Use mp_obj_malloc*() to allocate space for objects (aka structs with a type pointer) that will be
+// collected.
+
+#define m_new(type, num) ((type *)(m_malloc_helper(sizeof(type) * (num), M_MALLOC_RAISE_ERROR)))
+#define m_new_struct_with_collect(type, num) ((type *)(m_malloc_helper(sizeof(type) * (num), M_MALLOC_RAISE_ERROR | M_MALLOC_COLLECT)))
+#define m_new_maybe(type, num) ((type *)(m_malloc_helper(sizeof(type) * (num), 0)))
+#define m_new0(type, num) ((type *)(m_malloc_helper(sizeof(type) * (num), M_MALLOC_ENSURE_ZEROED | M_MALLOC_RAISE_ERROR)))
```

Comment:
As with the `malloc` family of macros above, I think it would be safer to require `m_new_with_collect()` and `m_new_without_collect()`, and similarly for related macros, and get rid of plain `m_new()`, etc., or else make the old names imply `_with_collect`.